internal/pkg/gateway/api.go

```diff
@@ -427,51 +428,111 @@ func (gs *Server) Submit(ctx context.Context, request *gp.SubmitRequest) (*gp.Su
 		return nil, status.Errorf(codes.Unavailable, "no orderer nodes available")
 	}
 
-	// try each orderer in random order
 	var errDetails []proto.Message
-	for _, index := range rand.Perm(len(orderers)) {
-		orderer := orderers[index]
-		logger.Infow("Sending transaction to orderer", "txID", request.TransactionId, "endpoint", orderer.logAddress)
-
-		var response *ab.BroadcastResponse
-		var err error
-		done := make(chan struct{})
-		go func() {
-			defer close(done)
-			ctx, cancel := context.WithTimeout(ctx, gs.options.BroadcastTimeout)
-			defer cancel()
-
-			response, err = gs.broadcast(ctx, orderer, txn)
-		}()
-		select {
-		case <-done:
-			// Broadcast completed normally
-		case <-ctx.Done():
-			// Overall submit timeout expired
-			logger.Warn("Submit call timed out while broadcasting to ordering service")
-			return nil, newRpcError(codes.DeadlineExceeded, "submit timeout expired while broadcasting to ordering service")
+	config := gs.getChannelConfig(request.ChannelId)
+	if config.ChannelConfig().Capabilities().ConsensusTypeBFT() {
+		// For BFT, we send transaction to ALL orderers
+		waitCh := make(chan *gp.ErrorDetail, len(orderers))
+		for _, o := range orderers {
+			go func(ord *orderer) {
+				logger.Infow("Sending transaction to orderer", "txID", request.TransactionId, "endpoint", ord.logAddress)
+				ctx, cancel := context.WithTimeout(ctx, gs.options.BroadcastTimeout)
+				defer cancel()
+				response, err := gs.broadcast(ctx, ord, txn)
+
+				if err != nil {
+					logger.Warnw("Error sending transaction to orderer", "txID", request.TransactionId, "endpoint", ord.logAddress, "err", err)
+					waitCh <- errorDetail(ord.endpointConfig, err.Error())
+				} else if status := response.GetStatus(); status == common.Status_SUCCESS {
+					logger.Infow("Successful response from orderer", "txID", request.TransactionId, "endpoint", ord.logAddress)
+					waitCh <- nil
+				} else {
+					logger.Warnw("Unsuccessful response sending transaction to orderer", "txID", request.TransactionId, "endpoint", ord.logAddress, "status", status, "info", response.GetInfo())
+					waitCh <- errorDetail(ord.endpointConfig, fmt.Sprintf("received unsuccessful response from orderer: %s", common.Status_name[int32(status)]))
+				}
+			}(o)
 		}
-
-		if err != nil {
-			errDetails = append(errDetails, errorDetail(orderer.endpointConfig, err.Error()))
-			logger.Warnw("Error sending transaction to orderer", "txID", request.TransactionId, "endpoint", orderer.logAddress, "err", err)
-			continue
+		n := len(orderers)
+		ordererConfig, ok := config.OrdererConfig()
+		if ok {
+			n = len(ordererConfig.Consenters())
 		}
-
-		status := response.GetStatus()
-		if status == common.Status_SUCCESS {
-			return &gp.SubmitResponse{}, nil
+		f := (float64(n) - 1) / 3
+		quorum := int(math.Ceil(2*f + 1))
```

Comment:
this is not how you compute the quorum :( 

I know I told you 2f+1, but, I always use this term because it's I usually imagine that `n` is congruent to 1 modulo 3. 

However, if we have, for example - `n=9` then `f=2` because we can only support 2 failures (to support 3 failures we need 10 nodes but we have 9). 

In that case, `2f+1` is 5, which means that if the leader is Byzantine, if you only get 5 success acknowledgements and 4 failures, the leader is free to ignore the transaction because 5 nodes cannot lobby for a view change in such a setting. 
However, if you require `n-f` which is 7, then the leader cannot ignore the request because it will eventually be overthrown by the 7 nodes (of course, assuming you actually succeeded sending to 7 honest nodes)

Please look below how to compute the quorum:

https://github.com/SmartBFT-Go/consensus/blob/v2/internal/bft/util.go#L169-L182

Reply:
I agree with Yakov, he's right.
Here's a chart you can use to write a test
| n  |  f | 2f+1 (n-f) |
|---|---|---|
| 1...3  | error  | error  |
| 4  | 1  | 3  |
| 5  | 1  | 4  |
| 6  | 1  | 5  |
| 7  | 2  | 5  |
| 8  | 2  | 6  |
| 9  | 2  | 7  |
| 10  | 3  | 7  |
| 11  | 3  | 8  |